src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles.SymbolSpecification;
+
+namespace Microsoft.CodeAnalysis.ConvertAutoPropertyToFullProperty
+{
+    internal abstract class AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider 
+        : CodeRefactoringProvider
+    {
+        internal abstract SyntaxNode GetPropertyDeclaration(SyntaxToken token);
+        internal abstract bool isAbstract(SyntaxNode property);
+        internal abstract bool TryGetEmptyAccessors(SyntaxNode propertyDeclarationSyntax, 
+            out SyntaxNode emptyGetAccessor, out SyntaxNode emptySetAccessor);
+        internal abstract SyntaxNode UpdateAccessor(SyntaxNode accessor, SyntaxNode[] statements);
+        internal abstract Task<SyntaxNode> ConvertToExpressionBodyIfDesiredAsync(Document document, 
+            SyntaxNode getAccessor, CancellationToken cancellationToken);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var property = GetPropertyDeclaration(token);
+            if (property == null)
+            {
+                return;
+            }
+
+            if (isAbstract(property))
+            {
+                return;
+            }
+
+            // check to see if both property accessors exist and are empty
+            if (!TryGetEmptyAccessors(property, out var emptyGetAccessor, out var emptySetAccessor))
```

Comment:
We should consider supporting getter-only autoprops too - with a readonly field, etc.